### PR TITLE
Fiber

### PR DESCRIPTION
**Pull request by iamdustan**
Early Preview. Testing my approach from https://github.com/iamdustan/tiny-react-renderer/pull/24.

~~Will update description later tonight or tomorrow.~~

To test:
* Clone `yarn install`
* `node run.js`
* open second tab, run `node_modules/.bin/react-devtools`
* enjoy.

---

This PR is intended to reimplement `react-blessed` on top of `ReactFiberReconciler`, the new (almost supported) renderer API. It is keeping the React 14.x `stack` reconciler support around for now by moving those files to `src/stack`. The shared utilities between stack and fiber have been moved to `src/shared`. The Fiber reconciler is entirely implemented right now in `src/fiber/fiber.js`. The other two files `src/fiber/index.js` and `src/fiber/devtools.js` exist to support [`react-devtools`](http://github.com/facebook/react-devtools) integration.

TODO:

**Examples (as test cases)**
- [x] render examples/animation.jsx
- [x] render examples/dashboard.jsx
- [x] render examples/demo.jsx

**Reconciler API**
- [x] `removeChild`
- [x] `insertBefore`
- [x] `*Text*` methods
- [x] clean up `screen.render()` calls
- [x] ensure event handlers work
- [ ] consider using async scheduling (fiber feature)
- [ ] consider supporting animation and deferred callbacks
- [ ] figure out how to publish/version (current requirement involves `npm install`ing an official renderer to get access to `react-{renderer}/lib/ReactFiberReconciler.js`

### CODE REVIEW

**Comment by iamdustan** [comment]
Current status:

react-blessed-devtools

**Comment by iamdustan** [comment]
Note that 2447 lines of the diff is yarn.lock. The primary fiber implementation is 182 lines of code.

**Comment by Yomguithereal** [comment]
That's really cool @iamdustan. Tell me when you are done with it and will merge. I guess we'll have to wait a bit for react 16 to be released though.

> figure out how to publish/version (current requirement involves npm installing an official renderer to get access to react-{renderer}/lib/ReactFiberReconciler.js

Should we engage a discussion with react team about this?

**Comment by iamdustan** [comment]
I’ve half-spoken to them about how to distribute the reconciler. There isn’t a clear direction yet and I don’t think it’s high-pri. I’m going to start another conversation in the [React Internals](https://www.messenger.com/t/1221515977927576) facebook group about it.

**Comment by iamdustan** [comment]
react-blessed-events

Added a couple more examples and have event handlers working. Honestly, idk if the last three items (async work) matters at this point. ReactDOM has it disabled for the initial fiber release and I think that’s okay here, too.

Regarding the publishing stuff, I’ll be opening an official issue on the React issue tracker later tonight or tomorrow. I think merging and publishing this as an alpha is okay (it should be feature compatible with previous releases, just on fiber and requiring a specific package of react and react-dom for the time being)

**Comment by Yomguithereal** [comment]
Thanks @iamdustan. I'll try to review the code thoroughly soon. I agree the most blocking point at that time is the fact that we'll need to rely on another renderer to get the necessary files to depend on. Just ping me on the issue you'll open please.

**Comment by Yomguithereal** [comment]
I am also open to modify some features of the library to "modernize" them in the process. The class system comes to mind :)

**Comment by iamdustan** [comment]
Issue for ReactFiberReconciler package opened at https://github.com/facebook/react/issues/9103.
Re: modernizing: I’ve started looking at integrating flexbox layout via [Yoga](https://github.com/facebook/yoga/). :)

**Comment by Yomguithereal** [comment]
Yes there was some idea once to adapt the CSS layout to the blessed styles. But I am not sure to see what it would entail.

**Comment by lapanoid** [comment]
Awesome stuff @iamdustan! I am playing with this currently and stuck with inputs.. Could you give a hint/example how input should work?

**Comment by lapanoid** [comment]
Awesome work @iamdustan ! I have played with your example, but failed with making inputs work having this alt tag. Tried on iterm2 and terminal. Is there a way to set focus won input without mouse involved, this is 90% use case for cli apps?

**Comment by Yomguithereal** [comment]
@lapanoid you should check `blessed`'s documentation about focus and inputs.

**Comment by iamdustan** [comment]
@lapanoid yeah, I’ve never had a lot of fun with blessed forms. I’ve had success with iterm2, but not with default Terminal.app

**Comment by Yomguithereal** [comment]
If I remember correctly, Terminal.app indeed has a lot of quirks, notably the mouse does not work and some focus events also. But it should be possible to devise a kind of form using tabulations & keys I guess.

**Comment by lapanoid** [comment]
@iamdustan My concern about react in terminal, that it should not break basic terminal functionality which are REPL and input.. All this gorgeous things will not get much traction if they are not play well with input.

**Comment by lapanoid** [comment]
@Yomguithereal sorry this is not right place for this discussion, I will create separate issue

**Comment by mashaal** [comment]
I had no luck using a `list` element -- errors out with `TypeError: Cannot read property 'length' of undefined`. Seems to work fine in previous version.

**Comment by iamdustan** [comment]
@mashaal could you post a brief demo and I’ll debug?

**Comment by mashaal** [comment]
@iamdustan https://gist.github.com/mashaal/2212d99f35dcadac2dfa096f7ae8623e

first example does render the list, second does not -- tried wrapping it in `element`, `box`, etc

**Comment by mitermayer** [comment]
Would love to see this PR progressing. What is its current state ?

**Comment by iamdustan** [comment]
Basically waiting on facebook/react#9103.

File: examples/animation.jsx [hunk 1]
```diff
@@ -1,43 +1,7 @@
-import React, {Component} from 'react';
+import React} from 'react';
```
Comment (devsnek) [inline examples/animation.jsx]: uhhhh?

**Comment by Yomguithereal** [comment]
Still waiting on Facebook :)

**Comment by Delapouite** [comment]
According to dan's answer, the missing piece should be released along React 16.1 in a week or two:

https://github.com/facebook/react/pull/10758#issuecomment-339154055

**Comment by Yomguithereal** [comment]
@iamdustan don't hesitate to comment on the thread when you feel we can start to work towards a merge.

**Comment by skellock** [comment]
This PR makes me happy in a way no PR should.

**Comment by iamdustan** [comment]
Will do @Yomguithereal. :smile:

Working through a few more issues (eg. facebook/react#11463) and then having you do an actual integration test somewhere would be really helpful.

**Comment by Yomguithereal** [comment]
> an actual integration test

Do you have an idea on how to do that with such a library?

**Comment by iamdustan** [comment]
Maybe I meant a manual test 😇 

I’ve added a few new examples, but I’m not sure if I broke the dashboard one or if it never really worked for me.

**Comment by iamdustan** [comment]
blessed

@Yomguithereal I think this is ready for a test drive.

**Comment by Delapouite** [comment]
Thanks ! I'll try to migrate https://github.com/byteclubfr/socket.io-monitor-cli over this new version to see how it goes.

File: package.json [hunk 2]
```diff
@@ -29,17 +29,24 @@
...
+    "babel-register": "^6.22.0",
     "blessed": "^0.1.81",
     "mocha": "^2.4.5",
-    "react": "^0.14.7"
+    "react": "16.1.0"
   },
   "dependencies": {
     "invariant": "^2.2.0",
-    "lodash": "^3.x.x"
+    "lodash": "^3.x.x",
```
Comment (Yomguithereal) [inline package.json]: I guess we can upgrade `lodash` to v4 now.

File: package.json [hunk 3]
```diff
@@ -29,17 +29,24 @@
...
+    "react-devtools": "^2.5.2",
+    "react-motion": "^0.5.2",
+    "react-reconciler": "0.5.0",
+    "ws": "^1.1.0",
+    "yoga-layout": "^1.2.0"
```
Comment (Yomguithereal) [inline package.json]: Shouldn't those dependencies be in `devDependencies`?

**Comment by Yomguithereal** [comment]
@iamdustan where are we on this? Doesn't some deps remain where they should not?

File: package.json (hunk 3 above)
Comment (iamdustan) [reply package.json]: Derp. Yoga-layout was added as part of an unfinished experiment so I just removed it. I’m not really sure if `ws` should be a devDep, peerDep, or regularDep. It’s required to use `react-devtools` which can be used in production mode (I’m pretty sure on that). Putting that as a devDep means consumers would have to add `ws` on their side, right?

**Comment by iamdustan** [comment]
I _think_ that this is ready. I removed one of the dependencies completely, but I think the other may be necessary still for `react-devtools` to work.

**Comment by jonhermansen** [comment]
Is this PR ready to be used by others? Are there any remaining tasks that I can help with to push this along?

**Comment by Yomguithereal** [comment]
@jonhermansen. This PR should be usable right now. I am just waiting an answer regarding dependencies to merge.

**Comment by iamdustan** [comment]
Derp. Missed the last question, @Yomguithereal. I just bumped a few versions and removed the dependency and added a guard during setup for it.

**Comment by Yomguithereal** [comment]
@iamdustan. I will merge & release now so that people can use your goodies. I'll try to find some time in the near future to work a bit on the library. Do you want me to add you as a contributor so you can freely edit the library?

**Comment by Yomguithereal** [comment]
@iamdustan:

> figure out how to publish/version (current requirement involves npm installing an official renderer to get access to react-{renderer}/lib/ReactFiberReconciler.js

Is this still the case?

**Comment by iamdustan** [comment]
Negative, the official renderer means that is no longer valid.

**Pull request merged**